std/experimental/testing/io.d

```diff
@@ -0,0 +1,326 @@
+
+import std.concurrency;
+import std.stdio;
+import std.conv;
+
+/**
+ * Write if debug output was enabled. Not thread-safe in the sense that it
+ * will get printed out immediately and may overlap with other output.
+ * This is why the test runner forces single-threaded mode when debug mode
+ * is selected.
+ */
+void writelnUt(T...)(T args)
+{
+    import std.stdio;
+
+    if (_debugOutput)
+        writeln("    ", args);
+}
+
+private shared(bool) _debugOutput = false; ///print debug msgs?
+private shared(bool) _forceEscCodes = false; ///use ANSI escape codes anyway?
+
+package void enableDebugOutput() nothrow
+{
+    synchronized
+    {
+        _debugOutput = true;
+    }
+}
+
+package bool isDebugOutputEnabled() nothrow
+{
+    synchronized
+    {
+        return _debugOutput;
+    }
+}
+
+package void forceEscCodes() nothrow
+{
+    synchronized
+    {
+        _forceEscCodes = true;
+    }
+}
+
+/**
+ * Adds to the test cases output so far or immediately prints
+ * Params:
+ *  output = The output to add to.
+ *  msg = The string to add.
+ */
+package void addToOutput(ref string output, in string msg) @safe
+{
+    if (_debugOutput)
+    {
+        import std.stdio;
+
+        writeln(msg);
+    }
+    else
+    {
+        output ~= msg;
+    }
+}
+
+package void utWrite(T...)(T args)
+{
+    WriterThread.get().write(args);
+}
+
+package void utWriteln(T...)(T args)
+{
+    WriterThread.get().writeln(args);
+}
+
+package void utWritelnGreen(T...)(T args)
+{
+    WriterThread.get().writelnGreen(args);
+}
+
+package void utWritelnRed(T...)(T args)
+{
+    WriterThread.get().writelnRed(args);
+}
+
+package void utWriteRed(T...)(T args)
+{
+    WriterThread.get().writeRed(args);
+}
+
+package void utWriteYellow(T...)(T args)
+{
+    WriterThread.get().writeYellow(args);
+}
+
+/**
+ * Thread to output to stdout
+ */
+class WriterThread
+{
+
+    /**
+     * Returns a reference to the only instance of this class.
+     */
+    static WriterThread get()
+    {
+        if (!_instantiated)
+        {
+            synchronized
+            {
+                if (_instance is null)
+                {
+                    _instance = new WriterThread;
+                }
+                _instantiated = true;
+            }
+        }
+        return _instance;
+    }
+
+    /**
+     * Writes the args in a thread-safe manner.
+     */
+    void write(T...)(T args)
+    {
+        _tid.send(text(args));
+    }
+
+    /**
+     * Writes the args in a thread-safe manner and appends a newline.
+     */
+    void writeln(T...)(T args)
+    {
+        write(args, "\n");
+    }
+
+    /**
+     * Writes the args in a thread-safe manner in green (POSIX only).
+     * and appends a newline.
+     */
+    void writelnGreen(T...)(T args)
+    {
+        _tid.send(green(text(args) ~ "\n"));
+    }
+
+    /**
+     * Writes the args in a thread-safe manner in red (POSIX only)
+     * and appends a newline.
+     */
+    void writelnRed(T...)(T args)
```

Comment:
What's the difference between 4 separate functions and 1 function with 4 possible instantiations?